File: 파이썬 알고리즘 인터뷰2/30. 중복 문자없는 가장 긴문자열.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters/

class Solution:
	def lengthOfLongestSubstring(self, s: str) -> int:
		start, end, maxSubstring = 0, 0, 0
		for c in s:
			if not c in s[start:end]:
				pass
			else:
				start = s[start:end].rindex(c) + start + 1
			end += 1
			maxSubstring = max(maxSubstring, end - start)
		return maxSubstring

File: problem/python algorithm interview/30. 중복 문자없는 가장 긴문자열.c
static char	*isUsed(char c, char	*start, char *end)
{
	while (start <= end)
	{
		end--;
		if (c == *end)
			return end;
	}
	return 0;
}

int lengthOfLongestSubstring(char * s){
	char	*start = s, *end = s;
	char	*tmp;
	size_t		maxSubstring = 0;

	while (*end)
	{
		tmp = isUsed(*end, start, end);
		if (!tmp)
			continue;
		else
			start = tmp + 1;
		end++;
		maxSubstring = maxSubstring > end - start ? maxSubstring : end - start;
	}
	return maxSubstring;
}
